### DISCUSSION TITLE
What's the reason behind "NewType cannot be used with structural type (a protocol or TypedDict class)"?

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pyright#8256] [Q&A] [answered]
I think that

```python
from typing import NewType, Protocol

class P(Protocol): ...

C = NewType("C", P)
```

is basically equivalent to

```python
from typing import Protocol

class P(Protocol): ...

class C(P):
    def __new__(cls, base: P, /) -> "C":
        return base  # type: ignore

    # TODO Forbid `isinstance` and `issubclass` usages.
```

. So why is the former snippet not allowed?

**Comment by erictraut** [comment]
You can't instantiate a protocol class, so `NewType` based on a protocol doesn't make sense.

Refer to [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/aliases.html#newtype) for details.

**Reply by Azureblade3808** [reply]
I don't see why protocol instantiation is required.

```python
from typing import NewType, Protocol

class P(Protocol): ...

C0 = NewType("C0", P)  # Let's assume this is allowed.

class C1(P): ...

c1 = C1()
c0 = C0(c1)
```
is how it would work. What am I missing?

**Reply by erictraut** [reply]
What are you trying to do here? How would you use the resulting type? Do you expect that it would be nominal or structural? If it's treated as structural, then it's equivalent to the original type, which defeats the purpose of creating a `NewType`. If it's nominal, then it's an abstract class because it doesn't implement the protocol, and it can't be used as a base class because a `NewType` class can't be used as a base class. I don't see any way that such a construct would be useful.

**Reply by Azureblade3808** [reply]
I expect the resulting type to be nominal.

> If it's nominal, then it's an abstract class because it doesn't implement the protocol,

It can be treated as concrete, because it can be "instantiated" (by returning the argument directly) and its every "instances" satisfies the base protocol.

**Reply by Azureblade3808** [reply]
I expect the resulting type to be nominal.

> If it's nominal, then it's an abstract class because it doesn't implement the protocol,

It can be treated as concrete, because it can be "instantiated" (by returning the argument directly) and its every "instances" satisfies the base protocol.

**Reply by erictraut** [reply]
But its "instances" do not satisfy anything but a trivial (empty) protocol definition (or a protocol that defines attributes and methods that are common to all objects).

If you want to define a nominal class that implements a protocol, `NewType` isn't the appropriate mechanism for this. A regular `class` statement is the appropriate mechanism.

If this is something you want to pursue further, I recommend taking the discussion to the [typing forum](https://discuss.python.org/c/typing/32).

**Reply by Azureblade3808** [reply]
I'm using `NewType` to distinguish different data of a same type.

The related code looks like --

```python
from typing import Any, NewType, Protocol

# region[from other modules]

class SomeProtocol(Protocol):
    # Some methods.
    def foo(self, /) -> Any: ...

def get_data_0() -> SomeProtocol: ...
def get_data_1() -> SomeProtocol: ...

# endregion

Data0 = NewType("Data0", SomeProtocol)
Data1 = NewType("Data1", SomeProtocol)

class Foo:
    @property
    def data_0(self, /) -> Data0:
        return Data0(get_data_0())

    @property
    def data_1(self, /) -> Data1:
        return Data1(get_data_1())

    @property
    def result(self, ) -> Any:
        return self.__compute(self.data_0, self.data_1)

    @staticmethod
    def __compute(data_0: Data0, data_1: Data1) -> Any: ...
```

It can be worked around if I define `Data0` and `Data1` as normal classes with corresponding special `__new__` methods and fake implementations of all protocol methods, but I think that defeats the very reason why `NewType` exists.

**Comment by Azureblade3808** [comment]
Also, I see some difference between Pyright and Mypy regarding NewType when its base is an ABC.

```python
from abc import ABC, abstractmethod
from typing import NewType, override

class Base(ABC):
    @abstractmethod
    def foo(self, /) -> int: ...

class Derived(Base):
    @override
    def foo(self, /) -> int:
        return 0

NewDerived = NewType("NewDerived", Base)

derived = Derived()
reveal_type(derived.foo)

new_derived = NewDerived(derived)  # ??? Should this be allowed?
reveal_type(new_derived.foo)
```

Mypy allows the line in question, whereas Pyright doesn't. I tend to agree with Mypy in this case.

**Reply by erictraut** [reply, marked as answer]
Yeah, I agree that an abstract nominal type should be allowed here. I'll fix this.

I'm still not convinced that a structural type should be allowed. If you want to pursue that idea, please post to the typing forum to get input from the broader community.

**Reply by Azureblade3808** [reply]
I'm looking forward for following code passing type-checks.

```
from typing import NewType, Protocol, override

class Base(Protocol):
    @abstractmethod
    def foo(self, /) -> int: ...

class Derived(Base):
    @override
    def foo(self, /) -> int:
        return 0

NewDerived = NewType("NewDerived", Base)  # type: ignore

derived = Derived()
reveal_type(derived.foo)

new_derived = NewDerived(derived)
reveal_type(new_derived.foo)
```

Currently, instantiation of `NewDerived` fails for the same reason ("reportAbstractUsage") with the former snippet. I hope they share the same logic and gets fixed at the same time. It will not be a burden to add some `type: ignore`s at NewType creations.
